packages/cardpay-sdk/sdk/reward-pool/base.ts

```diff
@@ -327,41 +264,55 @@ export default class RewardPool {
       return await waitForSubgraphIndexWithTxnReceipt(this.layer2Web3, txnHash);
     }
     let safeAddress = safeAddressOrTxnHash;
+    if (!proofArray) {
+      //proofArray can be empty e.g reward only for single rewardee
+      throw new Error('proof must be provided');
+    }
+    if (!leaf) {
+      throw new Error('leaf must be provided');
+    }
+
+    let { rewardProgramId, payee, token, amount }: FullLeaf = this.decodeLeaf(leaf) as FullLeaf;
+
     if (!rewardProgramId) {
       throw new Error('rewardProgramId must be provided');
     }
-    if (!tokenAddress) {
-      throw new Error('tokenAddress must be provided');
+    if (!leaf) {
+      throw new Error('leaf must be provided');
     }
-    if (!proof) {
-      throw new Error('proof must be provided');
+    if (!proofArray) {
+      throw new Error('proofArray must be provided');
+    }
+    if (!token) {
+      throw new Error('token must be provided');
+    }
+    if (!amount) {
+      throw new Error('amount must be provided');
     }
 
     let rewardManager = await getSDK('RewardManager', this.layer2Web3);
 
     if (!(await rewardManager.isRewardProgram(rewardProgramId))) {
       throw new Error('reward program does not exist');
     }
-
-    let from = contractOptions?.from ?? (await this.layer2Web3.eth.getAccounts())[0];
+    if (!(await rewardManager.isValidRewardSafe(safeAddress, rewardProgramId))) {
+      throw new Error('reward safe is not valid');
+    }
     let rewardSafeOwner = await rewardManager.getRewardSafeOwner(safeAddress);
-    let unclaimedRewards = new BN(await this.getBalanceForProof(rewardProgramId, tokenAddress, rewardSafeOwner, proof));
-    let rewardPoolBalanceForRewardProgram = (await this.balance(rewardProgramId, tokenAddress)).balance;
 
-    if (!(rewardSafeOwner == from)) {
-      throw new Error(
-        `Reward safe owner is NOT the signer of transaction.
-The owner of reward safe ${safeAddress} is ${rewardSafeOwner}, but the signer is ${from}`
-      );
-    }
-    let weiAmount = amount ? new BN(toWei(amount)) : unclaimedRewards;
-    if (weiAmount.gt(unclaimedRewards)) {
-      throw new Error(
-        `Insufficient rewards for rewardSafeOwner.
-For the proof, the reward safe owner can only redeem ${unclaimedRewards} but user is asking for ${amount}`
-      );
+    if (rewardSafeOwner != payee) {
+      throw new Error('payee is not owner of the reward safe');
     }
 
+    if (!(await this.isValid(leaf, proofArray))) {
+      throw new Error('proof is not valid');
+    }
+    if (!(await this.isClaimed(leaf))) {
+      throw new Error('reward has been claimed');
+    }
+    let from = contractOptions?.from ?? (await this.layer2Web3.eth.getAccounts())[0];
+    let weiAmount = new BN(amount);
+    let rewardPoolBalanceForRewardProgram = (await this.balance(rewardProgramId, token)).balance;
     if (weiAmount.gt(rewardPoolBalanceForRewardProgram)) {
```

Comment:
This doesn't deal properly with partial claims - there we expect the amount may be lower. Perhaps just wrapping with an extra if and adding a note in the exception that a workaround is to accept a partial claim

Reply:
I made a fix here